session/session.go

```diff
@@ -346,96 +295,143 @@ func (s *Session) handlePeriodicSearch(ctx context.Context) {
 }
 
 func (s *Session) randomLiveWant() cid.Cid {
-	if len(s.liveWants) == 0 {
+	s.sw.RLock()
+	defer s.sw.RUnlock()
+
+	if len(s.sw.liveWants) == 0 {
 		return cid.Cid{}
 	}
-	i := rand.Intn(len(s.liveWants))
+	i := rand.Intn(len(s.sw.liveWants))
```

Comment:
nit: Ideally, we'd avoid calling rand under a lock (call rand, get a rand int64, then use `%` under the lock).